File: app/src/main/java/com/nfcalarmclock/alarm/options/restrictvolume/NacRestrictVolumeDialog.java
package com.nfcalarmclock.alarm.options.restrictvolume;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.checkbox.MaterialCheckBox;

import com.nfcalarmclock.R;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.shared.NacSharedPreferences;
import com.nfcalarmclock.view.dialog.NacDialogFragment;

/**
 */
public class NacRestrictVolumeDialog
	extends NacDialogFragment
	implements View.OnClickListener
{

	/**
	 * Listener for when an audio source is selected.
	 */
	@SuppressWarnings("UnnecessaryInterfaceModifier")
	public interface OnRestrictVolumeListener
	{
		public void onRestrictVolume(boolean shouldRestrict);
	}

	/**
	 * Tag for the class.
	 */
	public static final String TAG = "NacRestrictVolumeDialog";

	/**
	 * Default volume restriction.
	 */
	private boolean mDefaultShouldRestrictVolume;

	/**
	 * Check box to restrict/unrestrict the volume.
	 */
	private MaterialCheckBox mShouldRestrictVolumeCheckBox;

	/**
	 * Summary text for whether volume should be restricted or not.
	 */
	private TextView mShouldRestrictVolumeSummary;

	/**
	 * Listener for when the volume is restricted/unrestricted.
	 */
	private OnRestrictVolumeListener mOnRestrictVolumeListener;

	/**
	 * Call the OnRestrictVolumeListener object, if it has been set.
	 */
	public void callOnRestrictVolumeListener()
	{
		OnRestrictVolumeListener listener = this.getOnRestrictVolumeListener();
		MaterialCheckBox checkBox = this.getShouldRestrictVolumeCheckBox();

		// Call the listener
		if ((listener != null) && (checkBox != null))
		{
			boolean isChecked = checkBox.isChecked();

			listener.onRestrictVolume(isChecked);
		}
	}

	/**
	 * Get the default volume restriction.
	 *
	 * @return The default volume restriction.
	 */
	public boolean getDefaultShouldRestrictVolume()
	{
		return this.mDefaultShouldRestrictVolume;
	}

	/**
	 * Get the OnRestrictVolumeListener object.
	 *
	 * @return The OnRestrictVolumeListener object.
	 */
	public OnRestrictVolumeListener getOnRestrictVolumeListener()
	{
		return this.mOnRestrictVolumeListener;
	}

	/**
	 * Get the restrict volume check box.
	 *
	 * @return The restrict volume check box.
	 */
	public MaterialCheckBox getShouldRestrictVolumeCheckBox()
	{
		return this.mShouldRestrictVolumeCheckBox;
	}

	/**
	 * Get the summary text for whether volume should be restricted or not.
	 *
	 * @return The summary text for whether volume should be restricted or not.
	 */
	private TextView getShouldRestrictVolumeSummary()
	{
		return this.mShouldRestrictVolumeSummary;
	}

	/**
	 * Called when the check box is clicked.
	 */
	@Override
	public void onClick(View view)
	{
		int id = view.getId();

		if (id == R.id.should_restrict_volume)
		{
			this.toggleShouldRestrictVolume();
			this.setupShouldRestrictVolumeSummary();
		}
	}

	/**
	 * Called when the dialog is created.
	 */
	@NonNull
	@Override
	public Dialog onCreateDialog(@Nullable Bundle savedInstanceState)
	{
		this.setupSharedPreferences();

		NacSharedConstants cons = this.getSharedConstants();

		return new AlertDialog.Builder(requireContext())
			.setTitle(cons.getTitleRestrictVolume())
			.setPositiveButton(cons.getActionOk(), (dialog, which) ->
				this.callOnRestrictVolumeListener())
			.setNegativeButton(cons.getActionCancel(), (dialog, which) -> {})
			.setView(R.layout.dlg_alarm_restrict_volume)
			.create();
	}

	/**
	 * Called when the fragment is resumed.
	 */
	@Override
	public void onResume()
	{
		super.onResume();

		AlertDialog dialog = (AlertDialog) getDialog();
		RelativeLayout container = dialog.findViewById(R.id.should_restrict_volume);

		// Set the member variable widgets
		this.mShouldRestrictVolumeCheckBox =
			dialog.findViewById(R.id.should_restrict_volume_checkbox);
		this.mShouldRestrictVolumeSummary =
			dialog.findViewById(R.id.should_restrict_volume_summary);

		container.setOnClickListener(this);
		this.setupShouldRestrictVolume();
		this.setupShouldRestrictVolumeColor();
	}

	/**
	 * Set the default volume restriction.
	 *
	 * @param  shouldRestrict  The default volume restriction.
	 */
	public void setDefaultShouldRestrictVolume(boolean shouldRestrict)
	{
		this.mDefaultShouldRestrictVolume = shouldRestrict;
	}

	/**
	 * Set the OnRestrictVolumeListener object.
	 *
	 * @param  listener  The OnRestrictVolumeListener object.
	 */
	public void setOnRestrictVolumeListener(
		OnRestrictVolumeListener listener)
	{
		this.mOnRestrictVolumeListener = listener;
	}

	/**
	 * Setup the check box and summary text for whether volume should be
	 * restricted or not.
	 */
	private void setupShouldRestrictVolume()
	{
		boolean shouldRestrict = this.getDefaultShouldRestrictVolume();

		this.getShouldRestrictVolumeCheckBox().setChecked(shouldRestrict);
		this.setupShouldRestrictVolumeSummary();
	}

	/**
	 * Setup the color of the restrict volume check box.
	 */
	private void setupShouldRestrictVolumeColor()
	{
		NacSharedPreferences shared = this.getSharedPreferences();

		int[] colors = new int[] { shared.getThemeColor(), Color.GRAY };
		int[][] states = new int[][] {
			new int[] {  android.R.attr.state_checked },
			new int[] { -android.R.attr.state_checked } };
		ColorStateList colorStateList = new ColorStateList(states, colors);

		this.getShouldRestrictVolumeCheckBox().setButtonTintList(colorStateList);
	}

	/**
	 * Setup the summary text for whether volume should be restricted or not.
	 */
	private void setupShouldRestrictVolumeSummary()
	{
		boolean shouldRestrict = this.shouldRestrictVolume();
		int textId = shouldRestrict ? R.string.restrict_volume_true
			: R.string.restrict_volume_false;

		this.getShouldRestrictVolumeSummary().setText(textId);
	}

	/**
	 * Get whether volume should be restricted or not.
	 *
	 * @return True if volume should be restricted, and False otherwise.
	 */
	private boolean shouldRestrictVolume()
	{
		return this.getShouldRestrictVolumeCheckBox().isChecked();
	}

	/**
	 * Toggle whether volume should be restricted or not.
	 */
	private void toggleShouldRestrictVolume()
	{
		boolean shouldRestrict = this.shouldRestrictVolume();

		this.getShouldRestrictVolumeCheckBox().setChecked(!shouldRestrict);
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/alarm/options/restrictvolume/NacRestrictVolumeDialog.kt
package com.nfcalarmclock.alarm.options.restrictvolume

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.navigation.fragment.findNavController
import com.google.android.material.button.MaterialButton
import com.google.android.material.checkbox.MaterialCheckBox
import com.nfcalarmclock.R
import com.nfcalarmclock.util.NacBundle
import com.nfcalarmclock.view.dialog.NacBottomSheetDialogFragment
import com.nfcalarmclock.view.setupCheckBoxColor

/**
 * Enable or disable the option to restrict the volume of an alarm that goes
 * off.
 */
class NacRestrictVolumeDialog
	: NacBottomSheetDialogFragment()
{

	/**
	 * Checkbox indicating whether to restrict volume or not.
	 */
	private lateinit var checkBox: MaterialCheckBox

	/**
	 * Called when the creating the view.
	 */
	override fun onCreateView(
		inflater: LayoutInflater,
		container: ViewGroup?,
		savedInstanceState: Bundle?
	): View?
	{
		return inflater.inflate(R.layout.dlg_restrict_volume, container, false)
	}

	/**
	 * Called when the view has been created.
	 */
	override fun onViewCreated(view: View, savedInstanceState: Bundle?)
	{
		// Super
		super.onViewCreated(view, savedInstanceState)

		// Get the bundle
		val alarm = NacBundle.getAlarm(arguments)

		// Get the views
		val okButton = dialog!!.findViewById(R.id.ok_button) as MaterialButton
		val cancelButton = dialog!!.findViewById(R.id.cancel_button) as MaterialButton

		// Get the default value
		val defaultRestrictVolume= alarm?.shouldRestrictVolume ?: false

		// Setup the views
		setupShouldRestrictVolume(defaultRestrictVolume)

		// Setup the ok button
		setupPrimaryButton(okButton, listener = {

			// Update the alarm attribute
			alarm?.shouldRestrictVolume = checkBox.isChecked

			// Save the change so that it is accessible in the previous dialog
			findNavController().previousBackStackEntry?.savedStateHandle?.set("YOYOYO", alarm)

			// Dismiss the dialog
			dismiss()

		})

		// Setup the cancel button
		setupSecondaryButton(cancelButton)
	}

	/**
	 * Setup whether to restrict volume or not.
	 */
	private fun setupShouldRestrictVolume(default: Boolean)
	{
		// Get the views
		val relativeLayout = dialog!!.findViewById(R.id.should_restrict_volume) as RelativeLayout
		val description = dialog!!.findViewById(R.id.should_restrict_volume_summary) as TextView
		checkBox = dialog!!.findViewById(R.id.should_restrict_volume_checkbox) as MaterialCheckBox

		// Set the default value
		checkBox.isChecked = default

		// Setup the views
		setupDescription(description)
		checkBox.setupCheckBoxColor(sharedPreferences)

		// Set the listener
		relativeLayout.setOnClickListener {

			// Toggle the checkbox
			checkBox.isChecked = !checkBox.isChecked

			// Setup the description
			setupDescription(description)

		}
	}

	/**
	 * Setup the text view for whether volume should be restricted or not.
	 */
	private fun setupDescription(textView: TextView)
	{
		// Determine the text ID to use based on whether restrict volume will
		// be used or not
		val textId = if (checkBox.isChecked)
		{
			R.string.restrict_volume_true
		}
		else
		{
			R.string.restrict_volume_false
		}

		// Set the text
		textView.setText(textId)
	}

	companion object
	{

		/**
		 * Tag for the class.
		 */
		const val TAG = "NacRestrictVolumeDialog"

	}

}